### PR TITLE
functionaltest: fix running tests under gdbserver

### PR DESCRIPTION
**Pull request by BK1603**
It was not possible to run the tests under the gdbserver because we were
not closing the old session before starting a new one. This caused the
server to not to be able to bind to the given address and crashing the
tests.

This commit closes the old session before starting a new one.

### CODE REVIEW

File: test/functional/helpers.lua [hunk 1]
```diff
@@ -410,6 +410,8 @@ end
 --    clear('-e')
 --    clear{args={'-e'}, args_rm={'-i'}, env={TERM=term}}
 function module.clear(...)
+  -- Close the old sesion before starting a new one.
+  module.set_session()
```
Comment (teto) [inline test/functional/helpers.lua:414]: `if session then session:close()  end` may make the intent clearer.

**Review by teto — APPROVED** [review]

File: test/functional/helpers.lua [hunk 2]
```diff
@@ -410,6 +410,10 @@ end
 --    clear('-e')
 --    clear{args={'-e'}, args_rm={'-i'}, env={TERM=term}}
 function module.clear(...)
+  -- Close the old sesion before starting a new one.
+  s = module.get_session()
```
Comment (justinmk) [inline test/functional/helpers.lua:414]: is there a reason local is not used? and why didn't our linter catch this?

```suggestion
  local s = module.get_session()
```

Comment (BK1603) [reply test/functional/helpers.lua]: No reason other than lack of sleep. Time for a cup of coffee I guess :)
Thank you for the review!

File: test/functional/helpers.lua [hunk 3]
```diff
@@ -410,6 +410,10 @@ end
 --    clear('-e')
 --    clear{args={'-e'}, args_rm={'-i'}, env={TERM=term}}
 function module.clear(...)
+  -- Close the old sesion before starting a new one.
```
Comment (justinmk) [inline test/functional/helpers.lua:413]:
```suggestion
  -- Close the old session before starting a new one.
```

**Review by justinmk — APPROVED** [review]

**Comment by BK1603** [comment]
@justinmk, I am not able to find where we close the session apart from in module.clear? Is this intended? If so, what happens to the session that is called after the last clear call?

**Comment by justinmk** [comment]
> I am not able to find where we close the session apart from in module.clear? Is this intended? If so, what happens to the session that is started in the last clear call?

Pretty just no one every bothered to look closely at it. I've wondered that myself a few times.

Thank you for looking into this. Eager to merge it, so if you have more improvements please followup in a new PR.

**Comment by justinmk** [comment]
oh interesting, this triggered asan. It's happening in asan, I wonder if it's related to https://github.com/neovim/neovim/blob/master/test/functional/plugin/lsp_spec.lua#L35-L39

```
021-09-19T17:27:40.4964711Z test/helpers.lua:249: Found runtime errors in logfile(s): 
...
2021-09-19T17:27:40.8036986Z 2 lua references were leaked!

2021-09-19T17:22:03.5678577Z [----------] Running tests from test/functional/plugin/lsp_spec.lua
...
2021-09-19T17:22:06.0589140Z [ RUN      ] LSP basic_init test should verify capabilities sent: 1301.42 ms OK
2021-09-19T17:22:06.1411835Z ==================== File /home/runner/work/neovim/neovim/build/log/asan.36195 ====================
2021-09-19T17:22:06.1412861Z = 
2021-09-19T17:22:06.1413427Z = =================================================================
2021-09-19T17:22:06.1414145Z = ==36195==ERROR: LeakSanitizer: detected memory leaks
2021-09-19T17:22:06.1414793Z = 
2021-09-19T17:22:06.1415569Z = Direct leak of 76 byte(s) in 1 object(s) allocated from:
2021-09-19T17:22:06.1416252Z =     #0 0x4d9f2d in malloc ??:?
2021-09-19T17:22:06.1417011Z =     #1 0x10f7ee2 in try_malloc /home/runner/work/neovim/neovim/src/nvim/memory.c:73:15
2021-09-19T17:22:06.1417898Z =     #2 0x10f80dc in xmalloc /home/runner/work/neovim/neovim/src/nvim/memory.c:107:15
2021-09-19T17:22:06.1418793Z =     #3 0x10f8676 in xmallocz /home/runner/work/neovim/neovim/src/nvim/memory.c:184:15
2021-09-19T17:22:06.1419772Z =     #4 0x10f87ce in xmemdupz /home/runner/work/neovim/neovim/src/nvim/memory.c:202:17
2021-09-19T17:22:06.1420680Z =     #5 0x10f9bd8 in xstrdup /home/runner/work/neovim/neovim/src/nvim/memory.c:421:10
2021-09-19T17:22:06.1421629Z =     #6 0xf6594f in nlua_luv_cfpcall /home/runner/work/neovim/neovim/src/nvim/lua/executor.c:240:5
2021-09-19T17:22:06.1424586Z =     #7 0x1cb9f85 in uv__run_timers /home/runner/nvim-deps/build/src/libuv/src/timer.c:174:5
2021-09-19T17:22:06.1425448Z = 
2021-09-19T17:22:06.1426178Z = SUMMARY: AddressSanitizer: 76 byte(s) leaked in 1 allocation(s).
2021-09-19T17:22:06.1426930Z 
```

**Comment by BK1603** [comment]
> Thank you for looking into this. Eager to merge it... need to figure out this ASAN failure first.
>

@justinmk, it's been bugging me too, fixing this should help me with other PRs. Hope I can help.
 
> It's happening in lsp_spec, I wonder if it's related to:
>

I would have to take a look at that, from what I've seen in the past hour, I don't know how that would affect LSP, and why would it not trigger for any tests other than this. 

What I can see from the stacktrace though, is that we try to execute a timer callback somewhere, which we fail to execute. The memory is then leaked from the following line (`multiqueue_put(main_loop.events, nlua_luv_error_event, 1, xstrdup(error))`):

https://github.com/neovim/neovim/blob/master/src/nvim/lua/executor.c#L240-#L241

We fail to free the memory we allocate in `xstrdup`, probably because we never get to call nlua_luv_error_event. The only way that I can think of my changes affecting this is by closing the session before we get to execute the timer callback, or the `nlua_luv_error_event` handler.

I don't know how LSP or these tests use libuv timers, I'll take a closer look at this and see what's going on.

**Comment by justinmk** [comment]
> We fail to free the memory we allocate in `xstrdup`, probably because we never get to call nlua_luv_error_event. The only way that I can think of my changes affecting this is by closing the session before we get to execute the timer callback, or the `nlua_luv_error_event` handler.

good hint! Sounds a lot like https://github.com/neovim/neovim/pull/15501 . Maybe we aren't draining Lua timer events before `nvim` exits ...?

**Comment by BK1603** [comment]
> good hint! Sounds a lot like #15501 . Maybe we aren't draining Lua timer events before `nvim` exits ...?

Could be that, but this line in main.c says that we iterate over all timers and close them. So I am not so sure about that, (unless maybe we miss adding the timer to the timers list?):

https://github.com/neovim/neovim/blob/master/src/nvim/main.c#L145
https://github.com/neovim/neovim/blob/853346a94d6aa78c97314a3b217fb5a5408a47f1/src/nvim/eval.c#L7878-#L7918

What I can say for sure for now is that we don't get to execute `nlua_luv_error_event`. So we are at least not draining these events from the main loop before exiting the session.

**Pull request merged**